### ISSUE TITLE
Unexpected Iteration Behavior with Strideable, ExpressibleByIntegerLiteral, and map

### ISSUE THREAD

**Issue by mohurwitzmusic** [apple/swift#65877]
**Description**
A discrepancy occurs in the Swift compiler's handling of Strideable and ExpressibleByIntegerLiteral protocols when mapping a range of integer literals to instances of a type that conforms to these protocols and clamps its values within a certain range. This leads to the production of an array with fewer elements than expected.

**Steps to reproduce**
Run the following code in a Playground:

```
import Foundation

struct Clamped: Strideable, ExpressibleByIntegerLiteral {
    
    var clamped: Int = 0 {
        didSet {
            clamped = min(max(clamped, 0), 10)
        }
    }
    
    func advanced(by n: Int.Stride) -> Clamped {
        return Clamped(clamped: clamped.advanced(by: n))
    }
    
    func distance(to other: Clamped) -> Int {
        return clamped.distance(to: other.clamped)
    }
    
    init(clamped: Int) {
        self.clamped = min(max(clamped, 0), 10)
    }
    
    init(integerLiteral value: Int) {
        self.clamped = min(max(value, 0), 10)
    }
}

enum Wrapped<T> {
    case wrapped(T)
}

var values: [Wrapped<Clamped>] = (0..<11).map { .wrapped($0) }
assert(values.count == 11) // 10
```

**Expected behavior**
The resultant array should have 11 elements.

Explicitly constructing the `Clamped` instances mitigates the probelem:

```
var values: [Wrapped<Clamped>] = (0..<11).map { .wrapped(Clamped(clamped: $0)) }
assert(values.count == 11) // passes
```

By adding globals to the top of the playground, further information can be attained:

```
import Foundation

var _ADVANCEMENTS_CALLED_ = 0
var _DISTANCES_CALLED_ = 0

struct Clamped {

    func advanced(by n: Int.Stride) -> Clamped {
        _ADVANCEMENTS_CALLED_ += 1
        return Clamped(clamped: clamped.advanced(by: n))
    }
    
    func distance(to other: Clamped) -> Int {
        _DISTANCES_CALLED_ += 1
        return clamped.distance(to: other.clamped)
    }
    
    ...
}
```

Now running the code using explicit construction produces the following output:

```
var values: [Wrapped<Clamped>] = (0..<11).map { .wrapped(Clamped(clamped: $0)) }
print(_ADVANCEMENTS_CALLED_) // 0
print(_DISTANCES_CALLED_) // 0
assert(values.count == 11) // 11
```

However running the code using the integer literal produces the following:

```
var values: [Wrapped<Clamped>] = (0..<11).map { .wrapped($0) }
print(_ADVANCEMENTS_CALLED_) // 10
print(_DISTANCES_CALLED_) // 53
assert(values.count == 11) // 10
```

 The problem appears to be that when these integer literals are mapped to Clamped instances using the .wrapped($0) notation, the Swift compiler implicitly uses the ExpressibleByIntegerLiteral initializer. This results in the clamping operation being performed, and since the clamping operation restricts the range to 0 to 10, the values 10 and 11 are treated as identical, hence the missing element in the resulting array.
 
 However, when you explicitly specify the type with .wrapped(Clamped(clamped: $0)), the compiler correctly uses the Clamped initializer that takes an Int parameter, not an integer literal. As such, each value in the range is treated as a distinct instance, and the resulting array has the correct number of elements.

 This discrepancy suggests that the Swift compiler might be making an unwarranted assumption about the semantics of ExpressibleByIntegerLiteral conformance when used within the map function. Ideally, each integer literal in the range should be considered a distinct value, regardless of how it's used to initialize a Clamped instance. The current behavior seems to contradict this, leading to unexpected results.

**Environment**
- swift-driver version: 1.62.15 Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51)
Target: arm64-apple-macosx13.0
- Xcode  14.2 (14C18)
- Deployment target: Xcode Playground

**Comment by tbkka** [comment]
I believe this is the expected behavior.  To see why, take a close look at this line:
```
var values: [Wrapped<Clamped>] = (0..<11).map { .wrapped($0) }
```

1. You've said the result is an array of `Wrapped<Clamped>`, so `.wrapped($0)` must produce that type
2. That implies that `$0` is of type `Clamped` 
3. So the range `(0..<11)` is `Range<Clamped>`
4. Which means that `0` and `11` here must be `Clamped` values

You would achieve exactly the same result if you had written this as:
```
var values = (Clamped(0)..<Clamped(11)).map { .wrapped($0) }
```

If you're familiar with C-style languages, Swift's handling of "integer literals" here is quite different.  In particular, an "integer literal" is _not an integer_.  Think of it as a piece of text inside the compiler that will get translated into some specific type _after_ the type checker has determined what the appropriate type is.  (That the translation occurs via an initializer `init(integerLiteral:)` that takes an `Int` argument is simply the practical way to implement this.)

In short, those literals `0` and `11` are `Clamped` values, not `Int` values.
